src/gui/attributetable/qgsattributetablefiltermodel.cpp

```diff
@@ -226,11 +226,18 @@ void QgsAttributeTableFilterModel::setSelectedOnTop( bool selectedOnTop )
   if ( mSelectedOnTop != selectedOnTop )
   {
     mSelectedOnTop = selectedOnTop;
+    int column = sortColumn();
+    Qt::SortOrder order = sortOrder();
 
-    if ( sortColumn() == -1 )
-    {
-      sort( 0 );
-    }
+    // set default sort values if they are not correctly set
```

Comment:
I leaved there to have more control on input data, as we saw that this methods can be called outside the qgis workflow (eg. in plugins).
E.g. in may part of QgsComposerAttributeTableModel the setSortOrder is set passing a value (EditRole) duing static_cast. That part of code it's not related with this one, but can explain how a user or qgis dev can wrongly setup sortOrder
Btw I'll move the input value control to QgsAttributeTableConfig::setSortOrder.